sphinx/source/docs/dev_guide/setup.rst

```diff
@@ -249,45 +278,144 @@ With either mode, you will be prompted for how to install BokehJS, e.g.:
 * ``python setup.py develop --install-js``
 
-Note that you will need to build BokehJS any time that the BokehJS source
-code changes (either by you or by pulling new revisions from GitHub). In
-particular, at the very least, you must build BokehJS the first time you
-install.
-
 .. note::
-    Occasionally the list of JavaScript dependencies also changes. If this
-    occurs, you will also need to re-run the instructions in the
-    :ref:`dev_guide_installing_node_packages` section above.
+    You need to rebuild BokehJS each time the BokehJS source code changes. This
+    can become necessary because you made changes yourself or because you pulled
+    updated code from GitHub.
+
+    Occasionally, the list of JavaScript dependencies also changes. If this
+    happens, you will need to re-run the instructions in the
+    :ref:`devguide_setup_installing_node_packages` section above before
+    rebuilding BokehJS.
 
-Downloading Sample Data
+    In case you update from a development environment based on Bokeh 2.3 or
+    older, you most likely also need to delete the ``bokehjs/build`` folder in
+    your local environment before building and installing a fresh BokehJS.
+
+.. _devguide_setup_sample_data:
+
+7. Download sample data
 -----------------------
 
-Several tests and examples require Bokeh's sample data to be available. Once
-Bokeh is installed, the simplest way to obtain the sample data is by executing
-the following command at a Bash or Windows prompt:
+Several tests and examples require Bokeh's sample data to be available on your
+hard drive. After :ref:`installing <devguide_setup_install_locally>` Bokeh, use
+the following command to download and install the data:
 
 .. code-block:: sh
 
     bokeh sampledata
 
-It's also possible to configure the download location, or to start the download
-programmatically. For full details see the :ref:`install_sampledata` section of
-the first steps guides.
+You also have the opportunity to configure the download location or to start the
+download programmatically. See the :ref:`install_sampledata` section of the
+first steps guides for more details.
+
+.. _devguide_setup_environment_variables:
+
+8. Set environment variables
+----------------------------
+
+Bokeh uses :ref:`environment variables <userguide_settings>` to control several
+aspects of how the different parts of the library operate and interact.
+
+To learn about all environment variables available in Bokeh, see
+:ref:`bokeh.settings` in the reference guide.
+
+``BOKEH_RESOURCES``
+~~~~~~~~~~~~~~~~~~~
+
+When working on Bokeh's codebase, the most important environment variable to be
+aware of is ``BOKEH_RESOURCES``. This variable controls which version of
+:term:`BokehJS` to use.
+
+By default, Bokeh downloads any necessary JavaScript code for BokehJS from a
+Content Delivery Network (CDN). If you want Bokeh to use your local BokehJS
+version instead, you should set ``BOKEH_RESOURCES`` to ``absolute-dev``:
```

Comment:
@hyles-lineata @bryevdv @mattpap This refers to the issue raised in #10694. Basically, the question is: What should we recommend for someone who generally wants to use a local BokehJS? When setting `BOKEH_DEV` to `true`, the variable ``BOKEH_RESOURCES`` will be set to ``absolute-dev``. This is why I assumed `absolute-dev` to be the best default value in this context.

However, I have mostly used `inline` in the past, same as Bryan. Matteusz recommends `server-dev`.

The paragraph above links to `bokeh.settings` which contains all information about these variables. I will also explain `BOKEH_RESOURCES` a little more in depth the upcoming BokehJS section of the dev guide, as Mateusz suggested.

Reply:
I don't think we are going to agree here on the advice here. So I'd actually just punt and be direct:

> By default, Bokeh loads latest published BokehJS from CDN. If you have modified and BokehJS and built it locally, you will need to make a configuration change in order to actually use the local version with your changes. Here are three ways to use local BokehJS:

And then explain all three.